code/ship/shiphit.cpp

```diff
@@ -2393,6 +2608,125 @@ static void ship_do_damage(object *ship_objp, object *other_obj, vec3d *hitpos,
+{
+	ship* shipp;
+	bool other_obj_is_weapon, other_obj_is_beam, other_obj_is_shockwave;
+
+	Assert(ship_objp);	// Goober5000
+	Assert(other_obj);
+	Assert(hitpos);		// Goober5000
+
+	Assert(ship_objp->instance >= 0);
+	Assert(ship_objp->type == OBJ_SHIP);
+	shipp = &Ships[ship_objp->instance];
+
+	Assert(other_obj->type == OBJ_WEAPON || other_obj->type == OBJ_BEAM || other_obj->type == OBJ_SHOCKWAVE);
+
+	// maybe adjust "damage" done by shockwave for BIG|HUGE
+	maybe_shockwave_damage_adjust(ship_objp, other_obj, &healing);
+
+	other_obj_is_weapon = ((other_obj->type == OBJ_WEAPON) && (other_obj->instance >= 0) && (other_obj->instance < MAX_WEAPONS));
+	other_obj_is_beam = ((other_obj->type == OBJ_BEAM) && (other_obj->instance >= 0) && (other_obj->instance < MAX_BEAMS));
+	other_obj_is_shockwave = ((other_obj->type == OBJ_SHOCKWAVE) && (other_obj->instance >= 0) && (other_obj->instance < MAX_SHOCKWAVES));
+	
+	MONITOR_INC(ShipHits, 1);
+
+	//	Don't heal player ship in the process of warping out.
+	if (Player->control_mode >= PCM_WARPOUT_STAGE2) {
+		if (ship_objp == Player_obj) {
+			return;
+		}
+	}
+
+	int wip_index = -1;
+	if (other_obj_is_weapon)
+		wip_index = Weapons[other_obj->instance].weapon_info_index;
+	else if (other_obj_is_beam)
+		wip_index = Beams[other_obj->instance].weapon_info_index;
+	else if (other_obj_is_shockwave)
+		wip_index = shockwave_get_weapon_index(other_obj->instance);
+
+	if (wip_index < 0)
+		return;
+	weapon_info* wip = &Weapon_info[wip_index];
+
+	// handle shield healing
+	if (!(ship_objp->flags[Object::Object_Flags::No_shields])) {
+		float shield_healing = healing * wip->shield_factor;
+
+		if (shield_healing > 0.0f) {
+
+			int dmg_type_idx = -1;
+
+			//get the "damage" type
+			if (other_obj_is_weapon || other_obj_is_beam) 
+				dmg_type_idx = wip->damage_type_idx;
+			else if (other_obj_is_shockwave) 
+				dmg_type_idx = shockwave_get_damage_type_idx(other_obj->instance);
+
+			if (shipp->shield_armor_type_idx != -1)
+				shield_healing = Armor_types[shipp->shield_armor_type_idx].GetDamage(shield_healing, dmg_type_idx, 1.0f, other_obj_is_beam);
+
+			shield_apply_healing(ship_objp, shield_healing);
+		}
+	}
+
+	// now for subsystems and hull
+	if ((healing > 0.0f)) {
+
+		do_subobj_heal_stuff(ship_objp, other_obj, hitpos, submodel_num, healing);
+
+		//Do armor stuff
+		int dmg_type_idx = -1;
+		if (other_obj_is_weapon || other_obj_is_beam)
+			dmg_type_idx = wip->damage_type_idx;
+		else if (other_obj_is_shockwave)
+			dmg_type_idx = shockwave_get_damage_type_idx(other_obj->instance);
+
+		if (shipp->armor_type_idx != -1)
+			healing = Armor_types[shipp->armor_type_idx].GetDamage(healing, dmg_type_idx, 1.0f, other_obj_is_beam);
+		
+		if (wip->wi_flags[Weapon::Info_Flags::Puncture])
+			healing /= 4;
+
+		healing *= wip->armor_factor;
+
+		// multiplayer clients don't do healing
+		if (!MULTIPLAYER_CLIENT) {
```

Comment:
Multiplayer clients shouldn't do *anything* healing related: shields, hull, or subsystem. So clients should probably abort much sooner than this.

Reply:
This is mostly copied from the structure of the damaging flow, should the damage *also* abort much sooner?

Reply:
Servers handle all actual damage calculation. Maybe in the damage codepath they already do abort sooner, and that was just someone thinking they are doing something when they are not.